_source/_posts/2022-02-23-angular-async-config.md

```diff
@@ -0,0 +1,348 @@
+export const environment = {
+  production: false,
+  authConfig: {
+    issuer: 'https://{yourOktaDomain}/oauth2/default',
+    clientId: '{yourClientId}'
+  }
+};
+```
+
+In `src/environments/environment.prod.ts`, you'll add the same `authConfig` properties with values that match your prod environment.
+
+You'll use the environment file to initialize the `OktaAuthModule` in the `AppModule` like this.
+
+```ts
+import { OktaAuthModule, OKTA_CONFIG } from '@okta/okta-angular';
+import { OktaAuth } from '@okta/okta-auth-js';
+import { environment } from '../environments/environment.ts';
+
+const oktaAuth = new OktaAuth({
+  ...environment.authConfig,
+  redirectUri: window.location.orgin + '/login/callback'
+});
+
+@NgModule({
+  declarations: [ AppComponent, ProfileComponent ],
+  imports: [ BrowserModule, AppRoutingModule, OktaAuthModule ],
+  providers: [
+    { provide: OKTA_CONFIG, useValue: { oktaAuth }}
+  ],
+  bootstrap: [AppComponent]
+})
+export class AppModule { }
+```
+
+**Summary of _configuration that changes by environment_:**
+
+Customizing environment files is the Angular recommended method to inject values during build time.
+
+**When to use:**
+
+* You have different configuration values based on build output
+
+**Best use cases:**
+
+* Devmode - keep apps served locally from doing things only prod apps should do 
+* Multiple staging environment systems
+
+**Watch out for:**
+
+* Configuration that involves private keys or tokens
+* Run the build for each environment to test changes you make. You don't want to miss adding a property and potentially get a runtime error.
+
+## Loading configurations from external APIs
+
+Sometimes you need to load configuration at runtime. This makes sense if you use release promotion style deployments - creating a build for a staging/pre-production environment and promoting the same build to production after verification. You don't want to create a new build, but what if your staging and production environments require different configurations? Loading configuration from an external API is handy in scenarios like these.
+
+To keep things simple for this *external API* configuration method, I'll focus only on the Okta example. 
+
+In this example, we'll look at `src/main.ts` where we bootstrap the Angular application. When you need configuration before the application loads, we can take advantage of `platformBrowserDynamic()` platform injector's `extraProviders` functionality. The `extraProviders` allows us to [provide platform providers](https://angular.io/guide/hierarchical-dependency-injection#platform-injector), much in the same way we can provide application-wide providers in the `AppModule`.
+
+Since we need to make the server call to get the configuration before we have a full Angular application context, we use Web APIs to call the API. Then we can configure the provider for Okta's `OKTA_CONFIG` injection token.
+
+For a configuration API call response that looks like this: 
+
+```json
```

Comment:
The syntax highlighting looks strange on this? Maybe it's because it wants everything in double-quotes?

```json
{
  "issuer": "https://{yourOktaDomain}/oauth2/default",
  "clientId": "{yourClientId}",
  "redirectUri": "{correctHostForTheEnvironment}/login/callback"
}

```

Reply:
Of course, makes perfect sense since I specified `json` syntax. Thanks for the catch!